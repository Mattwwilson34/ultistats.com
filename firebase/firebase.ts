// Import the functions you need from the SDKs you need
import { initializeApp } from 'firebase/app'
import { GoogleAuthProvider, getAuth } from 'firebase/auth'

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: 'AIzaSyBROS402FKa0UIKjziFiPgg9gir7DXNNrI',
  authDomain: 'ultistats-fc28e.firebaseapp.com',
  projectId: 'ultistats-fc28e',
  storageBucket: 'ultistats-fc28e.appspot.com',
  messagingSenderId: '289703592975',
  appId: '1:289703592975:web:23882caffc6e4527878f7a',
  measurementId: 'G-P6N3GKGE5X',
}

// Initialize Firebase
const app = initializeApp(firebaseConfig)

// Auth
const auth = getAuth(app)
const provider = new GoogleAuthProvider()

export { app, auth, provider }
