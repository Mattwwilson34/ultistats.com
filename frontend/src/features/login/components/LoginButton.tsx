import React, { useEffect, useState } from 'react'
import { Button, CircularProgress } from '@mui/material'
import { getRedirectResult, signInWithRedirect } from 'firebase/auth'
import { auth, provider } from '../../../../../firebase/firebase'
import checkUserExists from '../../../utils/check-user-exists'
import saveUserDataToDB from '../../../utils/save-user-to-db'

const LoginButton = (): JSX.Element => {
  const [redirected, setRedirected] = useState(false)

  useEffect(() => {
    const getUser = async (): Promise<void> => {
      try {
        const response = await getRedirectResult(auth)
        if (response == null) {
          throw Error('Response undefined')
        }

        const userData = response.user?.providerData[0]
        console.log(userData)
        if (userData === null || userData === undefined) {
          throw Error('User data undefined')
        }

        const userExistsInDB: boolean = await checkUserExists(userData)
        if (userExistsInDB) {
          localStorage.removeItem('redirectedFromGoogleAuth')
          window.location.href = '/'
        }

        await saveUserDataToDB(userData)
        localStorage.removeItem('redirectedFromGoogleAuth')
        window.location.href = '/'
      } catch (error) {
        console.error(error)
      }
    }
    const redirected = localStorage.getItem('redirectedFromGoogleAuth')
    if (redirected != null) {
      setRedirected((prevRedirected) => !prevRedirected)
      void getUser()
    }
  }, [])

  const handleLogin = (): void => {
    void (async () => {
      try {
        localStorage.setItem('redirectedFromGoogleAuth', 'true')
        await signInWithRedirect(auth, provider)
      } catch (error) {
        console.log(error)
      }
    })()
  }

  return (
    <>
      {!redirected ? (
        <Button variant="contained" size="large" onClick={handleLogin}>
          Login
        </Button>
      ) : (
        <CircularProgress />
      )}
    </>
  )
}

export default LoginButton
