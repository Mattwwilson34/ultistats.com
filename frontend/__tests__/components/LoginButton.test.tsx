/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  getRedirectResult,
  signInWithRedirect,
  getAuth,
  GoogleAuthProvider,
} from 'firebase/auth'
import React from 'react'
import LoginButton from '../../src/features/login'
import checkUserExists from '../../src/utils/check-user-exists'
import saveUserDataToDB from '../../src/utils/save-user-to-db'
import { auth } from '../../../firebase/firebase'
import 'jest-localstorage-mock'

jest.mock('../../../firebase/firebase')
jest.mock('../../src/utils/check-user-exists')
jest.mock('../../src/utils/save-user-to-db')
jest.mock('firebase/auth', () => ({
  getRedirectResult: jest.fn().mockResolvedValue({
    user: {
      providerData: [
        {
          providerId: 'google.com',
          uid: '116719161182905024316',
          displayName: 'Matt Wilson',
          email: 'mattwwilson34@gmail.com',
          phoneNumber: null,
          photoURL:
            'https://lh3.googleusercontent.com/a/AEdFTp4VE-2MhTCVIH5-kVZNziFheGMeDwJPpVG1X_WLpg=s96-c',
        },
      ],
    },
  }),
  signInWithRedirect: jest.fn(),
  getAuth: jest.fn(),
  GoogleAuthProvider: jest.fn(),
}))

describe('LoginButton', () => {
  beforeEach(() => {
    localStorage.clear()
  })

  it('renders a login button', () => {
    render(<LoginButton />)
    expect(screen.getByText('Login')).toBeInTheDocument()
  })

  it('does not show circular progress bar initially', () => {
    render(<LoginButton />)
    expect(screen.queryByRole('progressbar')).not.toBeInTheDocument()
  })

  it('shows circular progress bar when redirected from Google Auth', async () => {
    localStorage.setItem('redirectedFromGoogleAuth', 'true')

    render(<LoginButton />)

    expect(screen.queryByText('Login')).not.toBeInTheDocument()
    expect(screen.getByRole('progressbar')).toBeInTheDocument()
  })

  it('calls signInWithRedirect on login button click', async () => {
    render(<LoginButton />)
    const loginButton = screen.getByText('Login')
    await userEvent.click(loginButton)
    expect(signInWithRedirect).toHaveBeenCalled()
  })

  it('redirects to home page if user exists in DB', async () => {
    const providerData = [{ providerId: 'google.com', uid: '12345' }]
    const getRedirectResultMock = getRedirectResult as jest.MockedFunction<
      typeof getRedirectResult
    >
    getRedirectResultMock.mockResolvedValueOnce({ user: { providerData } })
    const checkUserExistsMock = checkUserExists as jest.MockedFunction<
      typeof checkUserExists
    >
    checkUserExistsMock.mockResolvedValueOnce(true)

    render(<LoginButton />)

    await waitFor(() => {
      expect(localStorage.getItem('redirectedFromGoogleAuth')).toBeNull()
      expect(window.location.href).toEqual('http://localhost/')
    })
  })

  it('saves user data to DB if user does not exist in DB', async () => {
    const providerData = {
      providerId: 'google.com',
      uid: '116719161182905024316',
      displayName: 'Matt Wilson',
      email: 'mattwwilson34@gmail.com',
      phoneNumber: null,
      photoURL:
        'https://lh3.googleusercontent.com/a/AEdFTp4VE-2MhTCVIH5-kVZNziFheGMeDwJPpVG1X_WLpg=s96-c',
    }

    const getRedirectResultMock = getRedirectResult as jest.MockedFunction<
      typeof getRedirectResult
    >
    getRedirectResultMock.mockResolvedValueOnce({ user: { providerData } })
    const checkUserExistsMock = checkUserExists as jest.MockedFunction<
      typeof checkUserExists
    >
    checkUserExistsMock.mockResolvedValueOnce(false)

    render(<LoginButton />)

    await waitFor(() => {
      expect(saveUserDataToDB).toHaveBeenCalledWith(providerData)
      expect(localStorage.getItem('redirectedFromGoogleAuth')).toBeNull()
      expect(window.location.href).toEqual('http://localhost/')
    })
  })
})
