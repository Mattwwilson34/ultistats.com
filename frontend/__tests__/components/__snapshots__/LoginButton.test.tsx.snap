// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginButton Component Renders LoginButtonComponent 1`] = `
<React.Fragment>
  <ForwardRef(Button)
    onClick={[Function]}
    size="large"
    variant="contained"
  >
    Login
  </ForwardRef(Button)>
</React.Fragment>
`;
